### ISSUE TITLE
[azservicebus] amqp-common's link.RetryableRPC() retries an entity does not exist

### ISSUE THREAD

**Issue by richardpark-msft** [Azure/azure-sdk-for-go#16088]
This can lead to a series of retries that will not work. It'd be nice to be able to pass in an error handling strategy per RPC call so we can provide better error handling when we know what the context of the call is and just fail early.

The underlying error is:

Code: 404
Description: "The messaging entity 'sb://<mysb>.servicebus.windows.net/samples' could not be found. To know more visit https://aka.ms/sbResourceMgrExceptions.  TrackingId:888b355f-038e-42e1-ab5d-cf1d0bfa20d1_G53, SystemTracker:riparkdev2.servicebus.windows.net:samples, Timestamp:2021-11-04T17:42:29"

**Comment by richardpark-msft** [comment]
Another example - when attempting to complete a message it will retry on 410 (message lock lost) which should be considered terminal and non-retryable.

**Comment by richardpark-msft** [comment]
@mikekistler noted that our retry policies just plain don't match the azcore retry numbers (and need to). This can cause some pretty bad retry behavior.

**Comment by vangent** [comment]
I'm in the process of trying to upgrade from the old "azure-service-bus-go" package, and I think this is also a change in behavior.

In the old package, when I try to send a message to a nonexistent topic, I would get an error back (not a great one, but at least it said "404" somewhere in it). Now, it just hangs until the context expires (if ever).

Same deal for trying to receive messages from a subscription that doesn't exist.

`SendMessage` and `ReceiveMessages` should return usable error messages for fatal errors, not just hang.

**Comment by richardpark-msft** [comment]
@vangent, I'll be working on this, this is definitely unintentional and is just us misclassifying errors error types.

**Comment by vangent** [comment]
Which module + version contains this fix?

**Comment by richardpark-msft** [comment]
Hi @vangent, 

You can install the latest release [v0.3.6](https://github.com/Azure/azure-sdk-for-go/releases/tag/sdk%2Fmessaging%2Fazservicebus%2Fv0.3.6) and get this fix and several others that impacted reliability and recoverability.

**Comment by vangent** [comment]
Thanks!
I can't find any documentation about error handling in the new library. The old one had constants like servicebus.ErrorNotImplemented, servicebus.ErrorNotFound, etc., but all I can find in the new one is 
https://github.com/Azure/azure-sdk-for-go/blob/sdk/messaging/azservicebus/v0.1.0/sdk/messaging/azservicebus/errors.go

Given that I made a call, and I got an error back, is there any structured way to determine what kind of error it is (other than parsing the message string, which fragile)?

**Comment by vangent** [comment]
BTW, I confirmed that I now do get an error back when sending to a nonexistent topic, although it takes > 60s to return. Maybe it's retrying internally for some time? Nonexistent topic seems like it should be non-retryable.

```
unhandled error link 4d9ac674-xxx: status code 404 and description: The messaging entity 'sb:/
/xxx.servicebus.windows.net/nonexistent-topic' could not be found. To know more visit https://aka.ms/sbResourceMgrException
```

**Comment by vangent** [comment]
Thanks! Any response to my question in https://github.com/Azure/azure-sdk-for-go/issues/16088#issuecomment-1062454958 ?

**Comment by richardpark-msft** [comment]
> Thanks! I can't find any documentation about error handling in the new library. The old one had constants like servicebus.ErrorNotImplemented, servicebus.ErrorNotFound, etc., but all I can find in the new one is https://github.com/Azure/azure-sdk-for-go/blob/sdk/messaging/azservicebus/v0.1.0/sdk/messaging/azservicebus/errors.go
> 
> Given that I made a call, and I got an error back, is there any structured way to determine what kind of error it is (other than parsing the message string, which fragile)?

We've tried to not redefine/recreate error types that exist, so for the most part the errors you'll see will be actual Go network related errors, or errors returned from go-amqp, which is our AMQP layer. It's not something we anticipate customers having to do, as most of the time an error returned from the library is a fatal one and there's no actual programmatic intervention possible.

We also allow you to configure the # and nature of the retries, so these can be adjusted if you find that the library is giving up too early or is taking too long between tries. You can configure this using the [ClientOptions](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus#ClientOption) when you create your client. This would let you still lean on the library for categorization of errors while still being more resilient. This could hopefully eliminate any need to manually categorize errors.

If you're still interested in dissecting the errors you can see how we categorization today using errors.As() our implementation is here: 
https://github.com/Azure/azure-sdk-for-go/blob/3349b7f7ed8c0af215b628b7734ebb2a7afdf876/sdk/messaging/azservicebus/internal/errors.go#L166
(looks like you've already looked at the older version of that file, it's seen some work since then)

It's totally possible to do the same in external code. We don't lock these errors down (behaviors in go-amqp could change) but you'll always see adjustments to azservicebus's error code to handle it.

All of this is a bit side to the original issue which is actually that the previous code doesn't use the error categorization code. I'm still fixing that one :)

**Comment by vangent** [comment]
OK, thanks. By the way, as I mentioned above I do get an error sending to a non-existent topic now (after 1m), but I still don't get anything but "context deadline exceeded" when trying to receive from a non-existent subscription (`servicebus.Receiver`.`ReceiveMessages`).

**Comment by richardpark-msft** [comment]
Yes, I'm definitely looking at that. I've got the test brewing and trying to see what I missed.

**Comment by jhendrixMSFT** [comment]
> as most of the time an error returned from the library is a fatal one and there's no actual programmatic intervention possible.

@richardpark-msft we should make this a strong guarantee (i.e. not "most" but "always") and document this behavior.

**Comment by richardpark-msft** [comment]
> Thanks! I can't find any documentation about error handling in the new library. The old one had constants like servicebus.ErrorNotImplemented, servicebus.ErrorNotFound, etc., but all I can find in the new one is https://github.com/Azure/azure-sdk-for-go/blob/sdk/messaging/azservicebus/v0.1.0/sdk/messaging/azservicebus/errors.go
> 
> Given that I made a call, and I got an error back, is there any structured way to determine what kind of error it is (other than parsing the message string, which fragile)?

ErrorNotFound is a good example of a useful errors - it'd be nice to at least provide some differentiation. I've filed https://github.com/Azure/azure-sdk-for-go/issues/17283 for further discussion.

**Comment by vangent** [comment]
@richardpark-msft  I still don't see this fixed. I'm using `v0.3.6`.

* Sending to a nonexistent topic fails, but it takes a long time still.
* Receiving (`ReceiveMessages`) from a nonexistent topic doesn't fail, it just waits for the context to timeout.

Can we keep the issue open until there's a confirmed fix in a tagged version?

**Comment by richardpark-msft** [comment]
Reopened. I even added a test for this so I must be missing some circumstance here. I'll look and see on Monday.

**Comment by richardpark-msft** [comment]
Oh, I see what you mean, yes, it'll be out in 0.3.7. I'll update this when it goes out.

**Comment by tkent** [comment]
@richardpark-msft I'm curious if this issue also explains some behavior we're seeing.

We've been seeing that a receiver of a topic that has a long period of inactivity will stop receiving new messages. It's extremely hard to reproduce on demand, but what we've seen is that a topic that receives no messages for a few days (2+) will not pickup new messages when they are added and the call will hang forever.

This comes up a lot with topics used for test environments where we may have a long period of inactivity.

We've worked around the issue by checking for messages in a loop with a 10 minute context timeout, but perhaps this a workaround we can back out once `0.3.7` is available?

**Comment by richardpark-msft** [comment]
hi @tkent, do you mind filing a separate issue for this? What you're mentioning is different than the original issue filed, and it's simpler to track yours separately.

**Comment by richardpark-msft** [comment]
@vangent , 0.4.0 has been released with this fix. Sorry about the earlyclose on that.

**Comment by vangent** [comment]
Works, thanks!

**Comment by tkent** [comment]
@richardpark-msft, we're going to let `v1.0.1` run for a few weeks without the work around we've had in place. If it shows up again, I'll file a separate issue. I'm hesitant to file an issue otherwise because we'll just get asked to try again with another version.

**Comment by tkent** [comment]
@richardpark-msft - I went through the exercise of running a small golang app using`v1.0.1` against a dedicated test service bus queue for about 13 days. When I put messages into the bus this evening, just as before, the app failed to receive new messages. I had to restart the app and then the messages showed up as expected.

I'll file an issue separately for this, but I'm going to put my workaround back in and call it day after that. I have to move onto other things.
